flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java

```diff
@@ -615,4 +632,15 @@ public static void runClusterEntrypoint(ClusterEntrypoint clusterEntrypoint) {
         /** Directly stops after the job has finished. */
         DETACHED
     }
+
+    /** The shutdown reason for {@link #shutDownAsync}. */
+    private enum ShutdownReason {
+        /**
+         * Flink needs to shutdown the cluster internally. We will deregister the application from
+         * resource management.
+         */
+        FLINK_INTERNAL,
+        /** The cluster entrypoint is killed externally. We will not deregister the application. */
+        EXTERNAL
+    }
```

Comment:
Rather than describing the shutdown reason I would recommend to describe the cleanup behaviour. Maybe something like

```
enum ShutdownBehaviour {
  STOP_APPLICATION,
  STOP_PROCESS,
}
```